pkg/controller/buildrun/runtime_image.go

```diff
@@ -0,0 +1,194 @@
+package buildrun
+
+import (
+	"bytes"
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"text/template"
+
+	buildv1alpha1 "github.com/redhat-developer/build/pkg/apis/build/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	// runtimeDockerfileTmpl Dockerfile template to be used with runtime-image, it uses Build attributes
+	// directly as template input.
+	runtimeDockerfileTmpl = `FROM {{ .Spec.Output.ImageURL }} as builder
+
+FROM {{ .Spec.Runtime.Base.ImageURL }}
+
+{{- range $k, $v := .Spec.Runtime.Env }}
+ENV {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $k, $v := .Spec.Runtime.Labels }}
+LABEL {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $dir := .Spec.Runtime.Directories }}
+{{- $parts := splitDirectories $dir }}
+COPY --from=builder "{{ index $parts 0 }}" "{{ index $parts 1 }}"
+{{- end }}
+
+{{- if .Spec.Runtime.WorkDir }}
+WORKDIR "{{ .Spec.Runtime.WorkDir }}"
+{{- end }}
+
+{{- if .Spec.Runtime.Entrypoint }}
+ENTRYPOINT [ {{ renderEntrypoint .Spec.Runtime.Entrypoint }} ]
+{{- end -}}
+`
+
+	// workspaceDir common workspace directory, where the source code is located.
+	workspaceDir = "/workspace/source"
+
+	// kanikoStable stable buildah image.
+	kanikoStable = "gcr.io/kaniko-project/executor:v0.23.0"
+
+	// runtimeDockerfile runtime Dockerfile file name.
+	runtimeDockerfile = "Dockerfile.runtime"
+)
+
+// rootUserID root's UID
+var rootUserID = int64(0)
+
+// runtimeDockerfilePath path to runtime Dockerfile on workspace directory.
+var runtimeDockerfilePath = path.Join(workspaceDir, runtimeDockerfile)
+
+// splitDirectories split informed directory by colon, returning a slice with parts. When colon is
+// not present, return the informed directory twice. This method always returns a slice with two
+// entries.
+func splitDirectories(dir string) []string {
+	parts := strings.Split(dir, ":")
+	if len(parts) == 2 {
+		return parts
+	}
+	return []string{dir, dir}
+}
```

Comment:
This should return `[]string, error`. The error should be returned if `len(parts) > 2`. You can also validate this in `build_controller.validateRuntime()`.

Reply:
This function is used inside the Go Templates ([FuncMap](https://golang.org/pkg/text/template/#FuncMap)) and in this context it's not possible to return error.

Reply:
Ok, good explanation. What do you think about adding validation code in `build_controller.validateRuntime()`. For each entry in `directories` (or `paths` once you rename it), `len(string.Split(path, ":"))` has to be either 1 or 2. Otherwise it is not value. Is that validation valuable?

Reply:
> Ok, good explanation. What do you think about adding validation code in `build_controller.validateRuntime()`. For each entry in `directories` (or `paths` once you rename it), `len(string.Split(path, ":"))` has to be either 1 or 2. Otherwise it is not value. Is that validation valuable?

I think would be simpler to have a single entry in `paths` (already renamed :-)), like for instance:

```yml
spec:
  runtime:
    paths:
      - /path/to/file                                         #  copying a single file
      - /path/to/directory                                    # copying a single directory
      - /path/to/another/directory:/target/location           # copying a directory with a new location
```

So the complexity of `.spec.runtime.paths` grows as the user needs more actions being taken. For a single file or directory, which should have the same location in the runtime-image uses a single entry.